### DISCUSSION TITLE
Communicate between viemodels in Avalonia

### DISCUSSION THREAD

**Discussion by vo1s** [AvaloniaUI/Avalonia#13878] [Q&A] [answered]
I know that there are many solutions for this issue in general. But i am new to this and couldnt implement it properly in AvaloniaUI for example with Pub/Sub pattern or DI. So please, show me the easiest way to get the property from from ViewModelA In ViewModelB.

**Comment by timunie** [comment]
well there is no "easiest" way. Ideas:
- if you use the MVVM Toolkit you can use a MessageBus
- You can pass a referece to the parent in the constructor of your ChildVM
- You can bind to DataContext of another named element if you want to access parent controls DataContext. 

To help further, please let us know which MVVM framework you use and what's the goal exactly is.

**Reply by vo1s** [reply]
Yes, I am using mvvm toolkit. I have a loginForm where the user login to the account. There are only two types of account: admin, user
Admin has every button enabled on the MainWindow, user has some of them Enabled. So I think I need to change the variable binded to IsEnabled of every needed buttons
```
[ObservableProperty] private bool _buttonsEnabled = false;
``` 
So for example if user logged in, I need to change ButtonsEnabled from MainViewModel to true in LoginViewModel. It means that I need an access to ButtonsEnabled that located in MainViewModel from LoginViewModel

**Reply by timunie** [reply]
I'd simply have `IsAdmin` in my MainViewModel and set this to true form the LoginVM. The LoginVM can have a reference to the MainVM. Now in you can either bind to commands and set CanExecute to false or bind IsEnabled to IsAdmin.

**Reply by vo1s** [reply]
So how can I set to true from loginVm? Can you show a quick example please.

**Reply by thevortexcloud** [reply]
The way I would do it would be to have the main view model instantiate the login view model. Then have it indirectly pass it to the view to create the login specific view. EG via an event or a [ReactiveUI interaction](https://docs.avaloniaui.net/docs/tutorials/music-store-app/opening-a-dialog). That way you don't break MVVM. The view is allowed to know about the view models, but the view models are not allowed to know about the view. View models can be aware of other view models if it makes sense to do so. In simple apps, having a sort of god class for your main view model is _probably_ fine. Although you could refactor that out to some sort of window service if you felt like it.

**Reply by timunie** [reply, marked as answer]
```cs
public partial class MainViewModel : ViewModelBase
{
    [ObservableProperty] 
    [NotifyCanExecuteChangedFor(nameof(DoThatAdminStuffCommand))]
    private bool _IsAdmin;
    
    [RelayCommand (CanExecute = nameof(IsAdmin))]
    void DoThatAdminStuff()
    {
        Debug.Print("I'm an Admin");
    }

    [RelayCommand]
    async Task LoginAsync()
    {
        var loginViewModel = new LoginViewModel();

        if (await loginViewModel.ShowAsync())
        {
            IsAdmin = loginViewModel.User.IsAdmin;
        }
    }
}

public class LoginViewModel : ViewModelBase
{
    public async Task<bool> ShowAsync()
    {
        // TODO: Show the LoginWindow
        // For ecample using a DialogManager (https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/ViewInteraction/DialogManagerSample) 
        await DialogManager.ShowLoginFor(this); // That is your task
        return loginWasOkay;
    }
}
```

Check Avalonia.Samples for ViewInteraction and Commands.

**Reply by timunie** [reply]
@thevortexcloud fyi, we have "newer" and extended samples for this kind of ViewInteraction :-)

**Reply by thevortexcloud** [reply]
Where can I find these questionably newer examples? Because I keep pointing people to the music store app. Which does work but it's also fairly complicated for someone who just wants a simple example for doing something like this.

EDIT: Found it https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/ViewInteraction/MvvmDialogSample#solution-2-write-your-own-interaction-class

**Reply by timunie** [reply]
and also this one: https://github.com/AvaloniaUI/Avalonia.Samples/tree/main/src/Avalonia.Samples/ViewInteraction/DialogManagerSample

needs some more "one time work" but is then as easy as `MyViewModel.ShowDialog()`
